### PR TITLE
Adventures in category theory - The algebra of types

### PR DESCRIPTION
**Pull request by miklos-martin**
Hey @itsdani and @andrasp3a,

Would you mind reviewing this?
The rendered html is here: https://github.com/miklos-martin/miklos-martin.github.io/blob/category-theory/_drafts/algebra-of-types.md

Thanks

### CODE REVIEW

File: _drafts/algebra-of-types.md [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+def unit[A]: A => Unit = _ => ()
+```
+
+Now let's take a look at some crazy ideas that our objects behave like numbers when we multiply them together with products or when we add them with sums (aka. coproduct).
+
+## Algebraic properties of products
+
+### Commutativity `a * b = b * a`
+
+So we ant to prove here, that the two scala types `A * B` and `B * A` are somehow equivalent. They are obviously not the same, but that is fine as long as they are isomorphic. What we need to do is swap the elements of a pair, it sounds fairly easy:
```
Comment (doczir) [inline _drafts/algebra-of-types.md]: "So we ant" => "So we want"

File: _drafts/algebra-of-types.md [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+proveTypesAreIsomorphic[String + (Int + Boolean), (String + Int) + Boolean].check
+// + OK, passed 100 tests.
+```
+
+### Unit `a + 0 = a`
+
+For product the unit was the terminal object and for its dual the unit is the terminal object's dual: the initial object.
+The reasoning is the same, just the arrows are reversed.
+
+![a + ø as sum][a + ø as sum]
```
Comment (doczir) [inline _drafts/algebra-of-types.md]: Image seem to be missing in the rendered version

File: _drafts/algebra-of-types.md [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+Where `ø` denotes the initial object. We want to prove that a is a better candidate for a coproduct than `a + ø` is.
+
+![a as coproduct][a as coproduct]
```
Comment (doczir) [inline _drafts/algebra-of-types.md]: Image seem to be missing in the rendered version

File: _drafts/algebra-of-types.md [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+
+Again, we can only use the universal construction, for which we have to find our injections and a morphism from `a` to `a + ø`. And they are, again, right there!
+
+![a is a better coproduct candidate][a is a better coproduct candidate]
```
Comment (doczir) [inline _drafts/algebra-of-types.md]: Image seem to be missing in the rendered version

File: _drafts/algebra-of-types.md [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+  def projectFst[A]: A * _ => A = _._1
+  def projectSnd[B]: _ * B => B = _._2
+}
+
+import Product._
+```
+
+## Coproduct
+
+In category theory, every construction has a dual, an inverse. If we invert the arrows in the definition of a product, we end up with the object `c` equipped with two _injections_ from `a` and `b`. Ranking two possible candidates is also inverted `c` is better than `c'` if there is a unique morphism from `c` to `c'` (so we could define `c'`'s injections by composition)
```
Comment (doczir) [inline _drafts/algebra-of-types.md]: What does it mean that `c` is better than `c'` in this context? How is it better? Some additional explanation might help.

File: _drafts/algebra-of-types.md [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+sealed trait Coproduct[A, B]
+case class One[A, B](a: A) extends Coproduct[A, B]
+case class Other[A, B](b: B) extends Coproduct[A, B]
+```
+
+Now we know why they are called algebraic.
+
+## Some extras
+
+Remember what was the first code snippet that was used to try the isomorphism in the [previous post]?
```
Comment (doczir) [inline _drafts/algebra-of-types.md]: Shouldn't this link work in the rendered version?

**Review by doczir — CHANGES_REQUESTED** [review]

File: _drafts/algebra-of-types.md (hunk 5 above)
Comment (miklos-martin) [reply _drafts/algebra-of-types.md]: it means the same as it meant in the discussion of universal constructions, or products, but I will add a few more thoughts to them as well.

File: _drafts/algebra-of-types.md (hunk 6 above)
Comment (miklos-martin) [reply _drafts/algebra-of-types.md]: I'm not sure, to be honest I have no idea why the images work, well, at least some of them :)

Comment (miklos-martin) [reply _drafts/algebra-of-types.md]: thank you for your comments

File: _drafts/algebra-of-types.md (hunk 5 above)
Comment (doczir) [reply _drafts/algebra-of-types.md]: Indeed, I only gave it a quick glance and manage to skip past that explanation. After a second read, the meaning is clear.

**Pull request merged**